linkerd/app/outbound/src/ingress.rs

```diff
@@ -63,106 +84,77 @@ impl Outbound<()> {
-            .instrument(|a: &tcp::Accept| info_span!("ingress", orig_dst = %a.orig_dst))
-            .push_map_target(|a: T| {
-                let orig_dst = Param::<OrigDstAddr>::param(&a);
-                tcp::Accept::from(orig_dst)
-            })
-            // Boxing is necessary purely to limit the link-time overhead of
-            // having enormous types.
-            .push(svc::BoxNewService::layer())
-            .check_new_service::<T, I>()
-            .into_inner()
+        http.push_on_response(
+            svc::layers()
+                .push(http::BoxRequest::layer())
+                .push(svc::MapErrLayer::new(Into::<Error>::into)),
+        )
+        // Lookup the profile for the outbound HTTP target, if appropriate.
+        //
+        // This service is buffered because it needs to initialize the profile
+        // resolution and a failfast is instrumented in case it becomes
+        // unavailable
+        // When this service is in failfast, ensure that we drive the
+        // inner service to readiness even if new requests aren't
+        // received.
+        .push_request_filter(http::Logical::try_from)
+        .check_new_service::<(Option<profiles::Receiver>, Target), _>()
+        .push(profiles::discover::layer(profiles, allow))
+        .push_on_response(
+            svc::layers()
+                .push(rt.metrics.stack.layer(stack_labels("http", "logical")))
+                .push(svc::layer::mk(svc::SpawnReady::new))
+                .push(svc::FailFast::layer("HTTP Logical", dispatch_timeout))
+                .push_spawn_buffer(buffer_capacity),
+        )
+        .push_cache(cache_max_idle_age)
+        .push_on_response(http::Retain::layer())
+        .instrument(|t: &Target| info_span!("target", dst = %t.dst))
+        // Obtain a new inner service for each request (fom the above cache).
+        //
+        // Note that the router service is always ready, so the `FailFast` layer
+        // need not use a `SpawnReady` to drive the service to ready.
+        .push(svc::NewRouter::layer(TargetPerRequest::accept))
+        .push(http::NewNormalizeUri::layer())
+        .push_on_response(
+            svc::layers()
+                .push(http::MarkAbsoluteForm::layer())
+                .push(svc::ConcurrencyLimit::layer(max_in_flight_requests))
+                .push(svc::FailFast::layer("HTTP Server", dispatch_timeout))
+                .push(rt.metrics.http_errors.clone())
+                .push(errors::layer())
+                .push(http_tracing::server(rt.span_sink, trace_labels()))
+                .push(http::BoxResponse::layer()),
+        )
+        .instrument(|a: &http::Accept| debug_span!("http", v = %a.protocol))
+        .push(http::NewServeHttp::layer(h2_settings, rt.drain))
+        .push_map_target(http::Accept::from)
+        .push(svc::UnwrapOr::layer(
+            svc::Fail::<_, IngressHttpOnly>::default(),
+        ))
```

Comment:
nit/TIOLI: is it necessary to use an `UnwrapOr` that produces a failing service here? seems like we could just use `push_request_filter`:
```suggestion
        .push_request_filter(|(version, accept)| { 
            Ok(version.ok_or(IngressHttpOnly)?, accept)
         })
```
not a big deal, just maybe slightly fewer moving parts.